File: pkg/chess/fen/fen.go
package fen

import (
	"math/bits"
	"strconv"
	"strings"

	"justchess/pkg/chess/bitboard"
	"justchess/pkg/chess/enums"
)

const DefaultFEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

func Bitboard2FEN(bb *bitboard.Bitboard) string {
	var fenStr strings.Builder

	fenStr.WriteString(serializePiecePlacement(bb.Pieces))
	fenStr.WriteString(serializeActiveColor(bb.ActiveColor))
	fenStr.WriteString(serializeCastlingRights(bb.CastlingRights))
	fenStr.WriteString(serializeEnPassantTarget(bb.EPTarget))
	fenStr.WriteString(strconv.Itoa(bb.HalfmoveCnt) + " ")
	fenStr.WriteString(strconv.Itoa(bb.FullmoveCnt))

	return fenStr.String()
}

var pieceSymbols = [12]byte{
	'P', 'p', 'N', 'n', 'B', 'b',
	'R', 'r', 'Q', 'q', 'K', 'k',
}

func serializePiecePlacement(bitboards [12]uint64) string {
	// Used to add characters to a string without extra mem allocs.
	var piecePlacementData strings.Builder

	var board [8][8]byte

	for pieceType, bitboard := range bitboards {
		// Go through all pieces on a bitboard.
		for ; bitboard > 0; bitboard &= bitboard - 1 {
			squareIndex := bits.TrailingZeros64(bitboard)
			// Add piece on board.
			board[squareIndex/8][squareIndex%8] = pieceSymbols[pieceType]
		}
	}

	var numOfEmptySquares byte

	for rank := 7; rank >= 0; rank-- {
		for file := 0; file < 8; file++ {
			char := board[rank][file]

			if char == 0 { // Empty square.
				numOfEmptySquares++
			} else { // Piece on square.
				if numOfEmptySquares > 0 {
					piecePlacementData.WriteByte('0' + numOfEmptySquares)
					numOfEmptySquares = 0
				}
				piecePlacementData.WriteByte(char)
			}

			// To add rank separators.
			squareIndex := 8*rank + file
			if (squareIndex+1)%8 == 0 {
				if numOfEmptySquares > 0 {
					piecePlacementData.WriteByte('0' + numOfEmptySquares)
					numOfEmptySquares = 0
				}
				// Do not add separator in the end of the string.
				if squareIndex != 7 {
					piecePlacementData.WriteByte('/')
				}
			}
		}
	}
	piecePlacementData.WriteByte(' ')

	return piecePlacementData.String()
}

func serializeActiveColor(c enums.Color) (ac string) {
	// White space is needed to be able to split the FEN fields.
	if c == enums.White {
		return "w "
	}
	return "b "
}

func serializeCastlingRights(cr [4]bool) string {
	var fenCRF strings.Builder

	if cr[0] {
		fenCRF.WriteByte('K')
	}
	if cr[2] {
		fenCRF.WriteByte('Q')
	}
	if cr[1] {
		fenCRF.WriteByte('k')
	}
	if cr[3] {
		fenCRF.WriteByte('q')
	}
	if fenCRF.Len() == 0 {
		fenCRF.WriteByte('-')
	}
	fenCRF.WriteByte(' ')

	return fenCRF.String()
}

func serializeEnPassantTarget(epTarget int) string {
	var fenEPF strings.Builder

	if epTarget < 0 {
		return "- "
	}
	// Calculate file and rank.
	files := "abcdefgh"
	fenEPF.WriteByte(files[epTarget%8])
	fenEPF.WriteString(strconv.Itoa((epTarget / 8) + 1))
	fenEPF.WriteByte(' ')

	return fenEPF.String()
}

func FEN2Bitboard(FEN string) *bitboard.Bitboard {
	fields := strings.Split(FEN, " ")
	pieces := parsePiecePlacement(fields[0])
	color := parseActiveColor(fields[1])
	castlingRights := parseCastlingRights(fields[2])
	epTarget := parseEnPassantTarget(fields[3])
	halfmoveClk, _ := strconv.Atoi(fields[4])
	fullmoveClk, _ := strconv.Atoi(fields[5])
	return bitboard.NewBitboard(pieces, color, castlingRights,
		epTarget, halfmoveClk, fullmoveClk)
}

// parsePiecePlacement parses bitboards from FEN's piece placement field.
func parsePiecePlacement(piecePlacementData string) [12]uint64 {
	var bitboards [12]uint64
	squareIndex := 56

	// Piece placement data describes each rank beginning from the eigth.
	for i := 0; i < len(piecePlacementData); i++ {
		char := piecePlacementData[i]

		if char == '/' { // Rank separator.
			squareIndex -= 16
		} else if char >= '1' && char <= '8' { // Number of consecutive empty squares.
			// Convert byte to the integer it represents.
			squareIndex += int(char - '0')
		} else { // There is piece on a square.
			var pieceType enums.PieceType // enum.PieceWPawn by default.
			// Manual switch construction is ~3x faster than map approach.
			switch char {
			case 'N':
				pieceType = enums.WhiteKnight
			case 'B':
				pieceType = enums.WhiteBishop
			case 'R':
				pieceType = enums.WhiteRook
			case 'Q':
				pieceType = enums.WhiteQueen
			case 'K':
				pieceType = enums.WhiteKing
			case 'p':
				pieceType = enums.BlackPawn
			case 'n':
				pieceType = enums.BlackKnight
			case 'b':
				pieceType = enums.BlackBishop
			case 'r':
				pieceType = enums.BlackRook
			case 'q':
				pieceType = enums.BlackQueen
			case 'k':
				pieceType = enums.BlackKing
			}
			// Set the bit on the bitboard to place a piece.
			bitboards[pieceType] |= 1 << squareIndex
			squareIndex++
		}
	}

	return bitboards
}

func parseActiveColor(fenACF string) enums.Color {
	if fenACF == "w" {
		return enums.White
	}
	return enums.Black
}

func parseCastlingRights(fenCRF string) (cr [4]bool) {
	for i := 0; i < len(fenCRF); i++ {
		switch fenCRF[i] {
		case 'K':
			cr[0] = true
		case 'k':
			cr[1] = true
		case 'Q':
			cr[2] = true
		case 'q':
			cr[3] = true
		}
	}
	return
}

func parseEnPassantTarget(fenEPF string) (index int) {
	if fenEPF == "-" {
		return -1
	}
	// Check file.
	switch fenEPF[0] {
	case 'a':
		index = 0
	case 'b':
		index = 1
	case 'c':
		index = 2
	case 'd':
		index = 3
	case 'e':
		index = 4
	case 'f':
		index = 5
	case 'g':
		index = 6
	case 'h':
		index = 7
	}
	rank := fenEPF[1] - '0'
	return index + int(rank-1)*8
}

File: _web/assets/scripts/chess/fen.js
import { Piece, Color, CastlingRights } from "/assets/scripts/chess/types.js"

export const initPos =
	"rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

// prettier-ignore
const square2String = [
	"a1", "b1", "c1", "d1", "e1", "f1", "g1", "h1",
	"a2", "b2", "c2", "d2", "e2", "f2", "g2", "h2",
	"a3", "b3", "c3", "d3", "e3", "f3", "g3", "h3",
	"a4", "b4", "c4", "d4", "e4", "f4", "g4", "h4",
	"a5", "b5", "c5", "d5", "e5", "f5", "g5", "h5",
	"a6", "b6", "c6", "d6", "e6", "f6", "g6", "h6",
	"a7", "b7", "c7", "d7", "e7", "f7", "g7", "h7",
	"a8", "b8", "c8", "d8", "e8", "f8", "g8", "h8",
]

/**
 * Parses piece the FEN string. It's the caller's responsibility to validate the input.
 * @param {string} fen
 * @returns {import("/assets/scripts/chess/types.js").Position}
 */
export function parseFen(fen) {
	const p =
		/** @type {import("/assets/scripts/chess/types.js").Position} */ ({})

	// Separate FEN fields.
	const fields = fen.split(" ", 6)

	// Parse piece placement.
	p.pieces = parsePiecePlacement(fields[0])

	// Parse active color.
	if (fields[1] === "b") {
		p.activeColor = Color.Black
	} else {
		p.activeColor = Color.White
	}

	// Parse castling rights.
	for (let i = 0; i < fields[2].length; i++) {
		switch (fields[2][i]) {
			case "K":
				p.castlingRights |= CastlingRights.WhiteShort
				break
			case "Q":
				p.castlingRights |= CastlingRights.WhiteLong
				break
			case "k":
				p.castlingRights |= CastlingRights.BlackShort
				break
			case "q":
				p.castlingRights |= CastlingRights.BlackLong
				break
		}
	}

	// Parse en passant target square.
	for (let i = 0; i < square2String.length; i++) {
		if (square2String[i] === fields[3]) {
			p.epTarget = i
		}
	}

	// Parse halfmove counter.
	p.halfmoveCnt = parseInt(fields[4])

	// Parse fullmove counter.
	p.fullmoveCnt = parseInt(fields[5])

	return p
}

/**
 * Parses piece the piece placement part of the FEN string.
 * It's the caller's responsibility to validate the input.
 * @param {string} piecePlacement
 * @returns {Map<number, Piece>}
 */
export function parsePiecePlacement(piecePlacement) {
	const pieces = new Map()

	let square = 56

	// Piece placement data describes each rank beginning from the eigth.
	for (let i = 0; i < piecePlacement.length; i++) {
		const char = piecePlacement[i]

		if (char === "/") {
			// Rank separator.
			square -= 16
			// Number of consecutive empty squares.
		} else if (char >= "1" && char <= "8") {
			square += parseInt(char)
		} else {
			// There is piece on a square.
			let piece = Piece.WPawn
			// Manual switch construction is ~3x faster than map approach.
			switch (char) {
				case "N":
					piece = Piece.WKnight
					break
				case "B":
					piece = Piece.WBishop
					break
				case "R":
					piece = Piece.WRook
					break
				case "Q":
					piece = Piece.WQueen
					break
				case "K":
					piece = Piece.WKing
					break
				case "p":
					piece = Piece.BPawn
					break
				case "n":
					piece = Piece.BKnight
					break
				case "b":
					piece = Piece.BBishop
					break
				case "r":
					piece = Piece.BRook
					break
				case "q":
					piece = Piece.BQueen
					break
				case "k":
					piece = Piece.BKing
					break
			}
			pieces.set(square, piece)
			square++
		}
	}

	return pieces
}
